### DISCUSSION TITLE
Data modelling question

### DISCUSSION THREAD

**Discussion by benjick** [Nozbe/WatermelonDB#1365] [Q&A]
Hello

I've been testing WatermelonDB a bit and decided to use it for my application. But I have some questions on how best to model the data.

On the top level my app will have "Projects", and under each project I will have different "Widgets"; for example TodoWidget and DateWidget. But I also want to be able to pull for example all TodoWidgets to create a summary for all things that needs to be done across all projects.

For example:
```
- Hiking 2022
- - TodoWidget: packlist
- - TodoWidget: book tickets
- - DateWidget: departure date
```

So how would I best model this? I have thought of two ways.

1. A table with `id`, `type` (todo, date, etc), `project_id`, and finally a `data` field which be a JSON field that the widget can modify freely.
2. One table for each type of widget with the defined schema for that widget.

I feel 2 would probably be better, but could it be a problem if I have too many relations between a project and all the different widgets? Would it also be possible to have a union type for a relation?

**Comment by ororsatti** [comment]
I think the first solution would be best, add the widget types you want and create parsers and type gaurds for them.
the second solution sounds a bit over the top for now, but maybe with scale you'd have to change.
